### ISSUE TITLE
property of variant in union type partially unknown

### ISSUE THREAD

**Issue by sbdchd** [microsoft/pyright#1841]
**Describe the bug**

Refining a union type based on the tag of the union results in the fields of the variants being partially `Unknown`.

**To Reproduce**

Running `pyright` on the code below produces the following messages:

```
main.py:21:21 - info: Type of "res.error" is "Unknown | str"
main.py:23:17 - info: Type of "res.code" is "str | Unknown"
```

**Expected behavior**

Ideally `Unknown` wouldn't be part of the type

**Screenshots or Code**
```python
from __future__ import annotations
from typing_extensions import Literal
from dataclasses import dataclass

@dataclass
class Success:
    ok: Literal[True]
    code: str

@dataclass
class Failure:
    ok: Literal[False]
    error: str

def get_auth_code() -> Success | Failure:
    ...

def main() -> None:
    res = get_auth_code()
    if not res.ok:
        reveal_type(res.error)
        return
    reveal_type(res.code)
```

**VS Code extension or command-line**
pyright 1.1.138

**Comment by erictraut** [comment]
Pyright contains support for type narrowing based on literal field types, but this works only for == and != operators. It doesn't work for general falsy or truthy tests (`if x` or `if not x`). You can make this work by changing your code to `if res.ok == False:`. 

I'm not inclined to expand the support for general falsy or truthy tests in this case because it adds complexity and a performance overhead, and I don't think this case is common enough to offset those.

**Comment by sbdchd** [comment]
Maybe this should be a separate ticket, but after seeing the truthiness check not work, the next thing I tried was:

```python
if res.ok is False: ...
``` 

which didn't refine it either

`if res.ok == False` works, but I'm accustomed to using `is` when comparing to singletons (`False`, `True`, `None`), probably due to pestering by `flake8`'s [`E711`](https://www.flake8rules.com/rules/E711.html) and `pylint`'s [`singleton-comparison`](http://pylint.pycqa.org/en/latest/technical_reference/features.html)

**Comment by erictraut** [comment]
If you want type narrowing to work in this case, you'll need to use `==` or `!=`. Here is [a complete list of the type guard idioms we support currently](https://github.com/microsoft/pyright/blob/master/docs/type-concepts.md#type-guards). Each type guard idiom requires added complexity and potentially has a perf impact on overall analysis performance, so we add support for new ones only when we have evidence that they are used commonly in Python code. 

I'm going to close this, but we can revisit it in the future if it receives enough upvotes or additional user requests.
